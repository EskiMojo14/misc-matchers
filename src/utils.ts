import type { MatcherHintOptions } from "jest-matcher-utils";
import type { MatcherUtils } from "expect";

// eslint-disable-next-line @typescript-eslint/no-explicit-any
export const isMock = (received: any): received is jest.Mock =>
  // eslint-disable-next-line @typescript-eslint/no-unsafe-member-access
  received?._isMockFunction === true;

// eslint-disable-next-line @typescript-eslint/no-explicit-any
export const isSpy = (received: any): received is jasmine.Spy =>
  // eslint-disable-next-line @typescript-eslint/no-unsafe-member-access
  typeof received?.calls?.all === "function" &&
  // eslint-disable-next-line @typescript-eslint/no-unsafe-member-access
  typeof received?.calls?.count === "function";

export function ensureMockOrSpy(
  utils: MatcherUtils["utils"],
  received: unknown,
  matcherName: string,
  expectedArgument?: string,
  options?: MatcherHintOptions,
): asserts received is jest.Mock | jasmine.Spy {
  if (!isMock(received) && !isSpy(received)) {
    throw new Error(
      utils.matcherErrorMessage(
        utils.matcherHint(matcherName, undefined, expectedArgument, options),
        `${utils.RECEIVED_COLOR(
          "received",
        )} value must be a mock or spy function`,
        utils.printWithType("Received", received, utils.printReceived),
      ),
    );
  }
}

/**
 * Given a label, return a function which given a string,
 * right-aligns it preceding the colon in the label.
 */
export const getRightAlignedPrinter = (label: string) => {
  // Assume that the label contains a colon.
  const index = label.indexOf(":");
  const suffix = label.slice(index);

  return (string: string, isExpectedCall: boolean) =>
    (isExpectedCall
      ? `->${" ".repeat(Math.max(0, index - 2 - string.length))}`
      : " ".repeat(Math.max(index - string.length))) +
    string +
    suffix;
};
