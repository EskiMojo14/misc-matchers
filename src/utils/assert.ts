export function assert(
  condition: unknown,
  message: string | (() => string),
): asserts condition {
  if (!condition) {
    throw new Error(typeof message === "function" ? message() : message);
  }
}
