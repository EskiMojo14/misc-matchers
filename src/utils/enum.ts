export interface EnumLike {
  [key: string]: string | number;
  [index: number]: string;
}

export const getValidEnumValues = (obj: EnumLike) =>
  Object.values(obj).filter((v) => typeof obj[v] !== "number");
