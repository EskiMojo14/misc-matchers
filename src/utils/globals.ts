// to avoid having separate jest and vitest suites, we create common definitions
// then mock during setup with the appropriate globals
import type {
  describe as jestDescribe,
  it as jestIt,
  expect as jestExpect,
  jest,
} from "@jest/globals";
import type {
  describe as vitestDescribe,
  it as vitestIt,
  expect as vitestExpect,
  vi,
  Mock as VitestMock,
} from "vitest";
import type { Satisfies } from "./types.typetest";

interface Each {
  <T extends Array<any> | [any]>(
    cases: ReadonlyArray<T>,
  ): (name: string, fn: (...args: T) => void) => void;
  <T>(cases: ReadonlyArray<T>): (name: string, fn: (value: T) => void) => void;
}
export declare const describe: (typeof vitestDescribe | typeof jestDescribe) & {
  each: Each;
};

// eslint-disable-next-line @typescript-eslint/no-unused-vars
type TestEach = Satisfies<
  typeof jestDescribe.each | typeof vitestDescribe.each,
  Each
>;

export declare const expect: typeof vitestExpect | typeof jestExpect;
export declare const it: typeof vitestIt | typeof jestIt;

export type Mock<R = any, A extends Array<any> = Array<any>, T = any> =
  | jest.Mock<(this: T, ...args: A) => R>
  | VitestMock<A, R>;

export declare const fn: {
  (): Mock;
  <R, A extends Array<any>, T>(fn?: (this: T, ...args: A) => R): Mock<R, A, T>;
};

// eslint-disable-next-line @typescript-eslint/no-unused-vars
type TestFn = Satisfies<typeof jest.fn | typeof vi.fn, typeof fn>;

throw new Error("This file should be mocked");
