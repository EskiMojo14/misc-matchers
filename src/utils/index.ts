import type { MatcherHintOptions } from "jest-matcher-utils";
import type { MatcherUtils } from "expect";

// eslint-disable-next-line @typescript-eslint/no-explicit-any
export const isMock = (received: any): received is jest.Mock =>
  // eslint-disable-next-line @typescript-eslint/no-unsafe-member-access
  received?._isMockFunction === true;

// eslint-disable-next-line @typescript-eslint/no-explicit-any
export const isSpy = (received: any): received is jasmine.Spy =>
  // eslint-disable-next-line @typescript-eslint/no-unsafe-member-access
  typeof received?.calls?.all === "function" &&
  // eslint-disable-next-line @typescript-eslint/no-unsafe-member-access
  typeof received?.calls?.count === "function";

export function ensureMockOrSpy(
  utils: MatcherUtils["utils"],
  received: unknown,
  matcherName: string,
  expectedArgument?: string,
  options?: MatcherHintOptions,
): asserts received is jest.Mock | jasmine.Spy {
  if (!isMock(received) && !isSpy(received)) {
    throw new Error(
      utils.matcherErrorMessage(
        utils.matcherHint(matcherName, undefined, expectedArgument, options),
        `${utils.RECEIVED_COLOR(
          "received",
        )} value must be a mock or spy function`,
        utils.printWithType("Received", received, utils.printReceived),
      ),
    );
  }
}
