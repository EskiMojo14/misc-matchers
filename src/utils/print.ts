/**
 * Given a label, return a function which given a string,
 * right-aligns it preceding the colon in the label.
 */
export const getRightAlignedPrinter = (label: string) => {
  // Assume that the label contains a colon.
  const index = label.indexOf(":");
  const suffix = label.slice(index);

  return (string: string, isExpectedCall: boolean) =>
    (isExpectedCall
      ? `->${" ".repeat(Math.max(0, index - 2 - string.length))}`
      : " ".repeat(Math.max(index - string.length))) +
    string +
    suffix;
};
