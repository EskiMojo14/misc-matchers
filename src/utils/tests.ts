/**
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

import ansiRegex from "ansi-regex";
// ansi-styles is ESM - as a result we need to run our Jest tests with `--experimental-vm-modules` flag
import style from "ansi-styles";
import type { NewPlugin } from "pretty-format";

export const alignedAnsiStyleSerializer: NewPlugin = {
  serialize(val: string | Error): string {
    // Return the string itself, not escaped nor enclosed in double quote marks.
    return (val instanceof Error ? val.message : val).replace(
      ansiRegex(),
      (match) => {
        switch (match) {
          case style.inverse.open:
            return "<i>";
          case style.inverse.close:
            return "</i>";

          case style.bold.open:
            return "<b>";
          case style.dim.open:
            return "<d>";
          case style.green.open:
            return "<g>";
          case style.red.open:
            return "<r>";
          case style.yellow.open:
            return "<y>";
          case style.bgYellow.open:
            return "<Y>";

          case style.bold.close:
            return "</b>";
          case style.dim.close:
            return "</d>";
          case style.green.close:
            return "</g>";
          case style.red.close:
            return "</r>";
          case style.yellow.close:
            return "</y>";
          case style.bgYellow.close:
            return "</Y>";

          default:
            return match; // unexpected escape sequence
        }
      },
    );
  },
  test(val: unknown) {
    return typeof val === "string" || val instanceof Error;
  },
};
