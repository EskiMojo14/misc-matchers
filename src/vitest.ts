/* eslint-disable @typescript-eslint/no-empty-interface */
import { expect } from "vitest";
import * as mixNMatchers from ".";

declare module "vitest" {
  interface Assertion<T> extends mixNMatchers.MixNMatchers<void, T> {}
  interface AsymmetricMatchersContaining
    extends mixNMatchers.AsymmetricMixNMatchers {
    enum: mixNMatchers.AsymmetricMixNMatchers["ofEnum"];
  }
}

expect.extend(mixNMatchers);
expect.extend({ enum: mixNMatchers.ofEnum });
