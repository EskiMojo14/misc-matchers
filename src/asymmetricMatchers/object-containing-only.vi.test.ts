import { expect, it, describe } from "vitest";
import { alignedAnsiStyleSerializer } from "../utils/tests";

expect.addSnapshotSerializer(alignedAnsiStyleSerializer);

describe("objectContainingOnly", () => {
  it("should check if object contains only the expected properties", () => {
    expect({ a: 1 }).toEqual(expect.objectContainingOnly({ a: 1, b: 2 }));

    expect(() => {
      expect({ a: 1, b: 2 }).toEqual(expect.objectContainingOnly({ a: 1 }));
    }).toThrowErrorMatchingSnapshot();

    expect(() => {
      expect({ a: 1, b: 2 }).toEqual(
        expect.objectContainingOnly({ a: 1, b: 3 }),
      );
    }).toThrowErrorMatchingSnapshot();
  });
  it("should throw an error if expected value is not an object", () => {
    expect(() => {
      expect({}).toEqual(expect.objectContainingOnly(1));
    }).toThrowErrorMatchingSnapshot();
  });
  it("should throw an error if received value is not an object", () => {
    expect(() => {
      expect(1).toEqual(expect.objectContainingOnly({}));
    }).toThrowErrorMatchingSnapshot();
  });
});
