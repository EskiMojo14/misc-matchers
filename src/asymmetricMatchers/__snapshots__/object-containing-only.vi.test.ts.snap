// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`objectContainingOnly > should check if object contains only the expected properties 1`] = `expected { a: 1, b: 2 } to deeply equal objectContainingOnly<[object Object]>`;

exports[`objectContainingOnly > should check if object contains only the expected properties 2`] = `expected { a: 1, b: 2 } to deeply equal objectContainingOnly<[object Object]>`;

exports[`objectContainingOnly > should check if object contains only the expected properties 3`] = `expected { a: 1, b: 2 } to deeply equal not.objectContainingOnly{â€¦}`;

exports[`objectContainingOnly > should throw an error if expected value is not an object 1`] = `Expected value must be an object, not number`;

exports[`objectContainingOnly > should throw an error if received value is not an object 1`] = `expected 1 to deeply equal objectContainingOnly<[object Object]>`;
